### ISSUE TITLE
[FeatureRequest] Support joined associations

### ISSUE THREAD

**Issue by raldred** [alecdotninja/active_record_distinct_on#22]
It's currently not possible to use joined associations with `distinct_on` and `order`
If you do, Postgres will error "SELECT DISTINCT ON expressions must match initial ORDER BY expressions"

So i tried to see if I can include the order expression as well, but it doesn't work...
Contrived example...

`Article.joins(:author).includes(:author).order(author: :id).distinct_on(:id, author: :id)`
This generates invalid SQL:
`SELECT DISTINCT ON ("articles"."{:author=>:id}")`

tried
`Article.joins(:author).includes(:author).distinct_on('"authors"."id"')`
This generates invalid SQL also
`SELECT DISTINCT ON ("articles"."""authors"".""id""")`

I realise this is a limitation of the gem, It would be nice if it could support this as currently it's not usable with any queries that order by a joins association.

I have added basic support for it in our fork.
https://github.com/residently/active_record_distinct_on/commit/f8c6dc5472000bb47ce7a030853f6c7ccba4bdb5

**Comment by alecdotninja** [comment]
I don't think accepting a hash for `distinct_on` is a good idea. It would make it inconsistent with other methods that accept columns (like `order` and `select`).

I have not personally tested this, but I think you should be able to write something like:

```ruby
Article.joins(:author).includes(:author).merge(Author.distinct_on(:id))
```

This strategy also works for `order` and `select`.

**Comment by raldred** [comment]
Fair point. I guess i figured given it's related to associations, i could follow a similar interface to` joins`, `includes`,` references` etc, which do take a variety of arguments.

I completely overlooked using `merge` here. I'll give that a whirl.

I do think if you want to remain consistent with select and order, `distinct_on` could take strings for associations eg. `authors.id`. particularly if you want to select distinct on an attribute on your primary resource as well as an associations attribute.

**Comment by alecdotninja** [comment]
> I do think if you want to remain consistent with select and order, distinct_on could take strings for associations

I agree! I am surprised that does not work. It seems like a bug to me.

**Comment by raldred** [comment]
Yeh it doesnt because the code only allows attributes on the current arel_table.

I've taken a crude approach to it here.
https://github.com/residently/active_record_distinct_on/commit/3fba30890a1995870f1391253706bade208d3466

This satisfies my requirement to use distinct on with a joined association.

**Comment by ezekg** [comment]
I'm open to accepting strings for association support if anybody wants to provide a PR, e.g.:

```rb
Author.joins(:books).distinct_on('books.title')
```

I've needed this before quite a few times.

**Comment by mijoharas** [comment]
Hey, we'd be keen for this feature too.

@raldred are you happy to raise your change from your forked branch as a PR?

@alecdotninja would you be keen to accept this if it were raised like that (strings allowing you to specify the table)?

**Comment by ezekg** [comment]
@mijoharas I'm one of the maintainers now too, happy to accept a PR here!

**Comment by raldred** [comment]
Happy to raise a PR, I'll take a look, I can't remember why I dropped the condition around a Hash, maybe it wasn't needed.

**Comment by mijoharas** [comment]
Oh awesome, nice stuff both! (sorry @ezekg , hadn't noticed).

@raldred I think it looks like you first went for a `other_joined_table: :column_name` syntax instead of `other_joined_table.column_name` in this commit https://github.com/residently/active_record_distinct_on/commit/f8c6dc5472000bb47ce7a030853f6c7ccba4bdb5 and then reverted back to what was on main to switch to the string version after that was suggested here in the next commit. (I'd probably squash both together and then merge 🤷‍♂ )

**Comment by raldred** [comment]
@mijoharas https://github.com/alecdotninja/active_record_distinct_on/pull/33
I decided to have it support formats, Hash or String
`other_joined_table: :column_name` and `other_joined_table.column_name`

**Comment by ezekg** [comment]
Released in v1.9.0! ty @raldred

**Comment by mijoharas** [comment]
Thank you both so much! (fyi, we've been testing this, and all looks good so far) 🙌

**Comment by raldred** [comment]
You're very welcome. I'm glad it helps.
